core/src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -232,6 +233,16 @@ public CompiledScript compile(Script script, ScriptContext scriptContext) {
         if (canExecuteScript(lang, scriptEngineService, script.getType(), scriptContext) == false) {
             throw new ScriptException("scripts of type [" + script.getType() + "], operation [" + scriptContext.getKey() + "] and lang [" + lang + "] are disabled");
         }
+
+        // special exception to prevent expressions from compiling as update or mapping scripts
+        boolean expression = scriptEngineService instanceof ExpressionScriptEngineService;
```

Comment:
It might be a good idea (possibly in a different PR) to have a method on `ScriptEngineService` called something like `getSupportedScriptContexts()` which each implementation can use to define what script APIs they support. I imagine there are/will be other language that don't support some script APIs and this would not only allow them to use this too but would also remove language specific code form the ScriptService, which should probably remain language agnostic.